src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,940 @@
+
+   if sym === :mat_iso || sym === :ring_iso
+      if isdefined(G,sym) return getfield(G,sym) end
+      fm = gen_mat_iso(G.deg, G.ring)
+      G.ring_iso=fm.fr
+      G.mat_iso=fm
+
+   elseif sym === :X
+      if isdefined(G,:X) return getfield(G,:X) end
+      if isdefined(G,:descr)
+         if !isdefined(G,:ring_iso)
+            fm = gen_mat_iso(G.deg, G.ring)
+            G.ring_iso=fm.fr
+            G.mat_iso=fm
+         end
+         assign_from_description(G)
+      elseif isdefined(G,:gens)
+         if !isdefined(G,:ring_iso)
+            fm = gen_mat_iso(G.deg, G.ring)
+            G.ring_iso=fm.fr
+            G.mat_iso=fm
+         end
+         V = GAP.julia_to_gap([g.X for g in G.gens])
+         G.X=GAP.Globals.Group(V)
+      end
+
+   elseif sym === :gens
+      if isdefined(G, :gens) return getfield(G, :gens) end
+      if !isdefined(G, :X)
+         if !isdefined(G,:ring_iso)
+            fm = gen_mat_iso(G.deg, G.ring)
+            G.ring_iso=fm.fr
+            G.mat_iso=fm
+         end
+         assign_from_description(G)
+      end
+      L = GAP.Globals.GeneratorsOfGroup(getfield(G,:X))
+      G.gens=[MatrixGroupElem(G,G.mat_iso(L[i]),L[i]) for i in 1:length(L)]
+
+   else
+      return getfield(G, sym)
+
+   end
+
+end
+
+
+# add field :X
+function Base.getproperty(x::MatrixGroupElem, sym::Symbol)
+   if sym === :X && !isdefined(x,:X)
+      if !isdefined(x.parent,:ring_iso)
+         fm = gen_mat_iso(x.parent.deg, x.parent.ring)
+         setfield!(x.parent,:ring_iso,fm.fr)
+         setfield!(x.parent,:mat_iso,fm)
+      end
+      setfield!(x, :X, x.parent.mat_iso(x.elm))
+   elseif sym == :elm && !isdefined(x, :elm)                        # assuming that x.X, hence x.parent.ring_iso, are defined
+      setfield!(x, :elm, x.parent.mat_iso(x.X))
+   end
+   return getfield(x,sym)
+end
+
+Base.IteratorSize(::Type{<:MatrixGroup}) = Base.SizeUnknown()
+
+function Base.iterate(G::MatrixGroup)
+  L=GAP.Globals.Iterator(G.X)
+  if GAP.Globals.IsDoneIterator(L)
+    return nothing
+  end
+  i = GAP.Globals.NextIterator(L)
+  return MatrixGroupElem(G, G.mat_iso(i),i), L
+end
+
+function Base.iterate(G::MatrixGroup, state)
+  if GAP.Globals.IsDoneIterator(state)
+    return nothing
+  end
+  i = GAP.Globals.NextIterator(state)
+  return MatrixGroupElem(G, G.mat_iso(i),i), state
+end
+
+########################################################################
+#
+# Membership
+#
+########################################################################
+
+
+function ==(G::MatrixGroup,H::MatrixGroup)
+   G.deg==H.deg || return false
+   G.ring==H.ring || return false
+   if isdefined(G, :descr) && isdefined(H, :descr)
+      return G.descr == H.descr
+   end
+   if isdefined(G, :gens) && isdefined(H, :gens)
+      if G.gens==H.gens return true  end       #TODO this works only if the == between elements does not keep track of the parent
+   end
+   return G.X==H.X
+end
+
+
+# this saves the value of x.X
```

Comment:
The embedding functions such as `G(x)` (for `G` group and `x` element/matrix) return a `MatrixGroupElem` value, but the underlying GAP object `x.X`, if required, is computed in the function `lies_in` (called by the embedding function). Hence, to avoid recomputing `x.X`, the function `lies_in` returns the value `x.X` as second output in order to be used in the embedding function. But I don't even know if just delete that line, since not it sound to me a quite obvious fact.